lib/fuser_mqubit.h

```diff
@@ -0,0 +1,1087 @@
+          }
+
+          for (auto gate : ogate2->gates) {
+            ogate1->gates.push_back(gate);
+          }
+        }
+
+        if (cur_size == max_fused_size) {
+          break;
+        }
+      }
+
+      FuseNext(1, *ogate1);
+
+      std::sort(ogate1->qubits.begin(), ogate1->qubits.end());
+
+      fused_gates.push_back({ogate1->parent->kind, ogate1->parent->time,
+                             std::move(ogate1->qubits), ogate1->parent,
+                             std::move(ogate1->gates)});
+
+      ++stat.num_fused_gates;
+    }
+  }
+
+  static void MakeGateSequence(
+      unsigned max_fused_size, Scratch& scratch, GateF& fgate) {
+    unsigned level = 2 + scratch.count;
+
+    FindLongestGateSequence(max_fused_size, level, scratch, fgate);
+
+    auto longest_seq = scratch.longest_seq;
+
+    if (longest_seq.size() == 1 && scratch.count == 0) {
+      fgate.visited = 1;
+      return;
+    }
+
+    ++scratch.count;
+
+    for (auto p : longest_seq) {
+      p->gate->visited = 900;
+
+      for (auto q : p->qubits) {
+        fgate.qubits.push_back(q);
+        fgate.mask |= uint64_t{1} << q;
+      }
+
+      for (auto l : p->links) {
+        fgate.links.push_back(l);
+      }
+    }
+
+    // Compress links.
+    for (auto& link : fgate.links) {
+      while (link->prev != nullptr && link->prev->val->visited == 900) {
+        link = link->prev;
+      }
+
+      while (link->next != nullptr && link->next->val->visited == 900) {
+        LinkManager::Delete(link->next);
+      }
+    }
+
+    for (auto p : longest_seq) {
+      p->gate->visited = level;
+    }
+
+    if (longest_seq.size() >= 3) {
+      AddGatesFromNext(longest_seq[2]->gate->gates, fgate);
+    }
+
+    if (longest_seq.size() >= 5) {
+      AddGatesFromNext(longest_seq[4]->gate->gates, fgate);
+    }
+
+    if (longest_seq.size() >= 2) {
+      // May call MakeGateSequence recursively.
+      AddGatesFromPrev(max_fused_size, *longest_seq[1]->gate, scratch, fgate);
+    }
+
+    if (longest_seq.size() >= 4) {
+      // May call MakeGateSequence recursively.
+      AddGatesFromPrev(max_fused_size, *longest_seq[3]->gate, scratch, fgate);
+    }
+
+    for (auto p : longest_seq) {
+      p->gate->visited = 999;
+    }
+
+    FuseNext(1, fgate);
+
+    scratch.gates.push_back(&fgate);
+  }
+
+  static void AddGatesFromNext(std::vector<const Gate*>& gates, GateF& fgate) {
+    for (auto gate : gates) {
+      fgate.gates.push_back(gate);
+    }
+  }
+
+  static void AddGatesFromPrev(unsigned max_fused_size, const GateF& pfgate,
+                               Scratch& scratch, GateF& fgate) {
+    for (auto gate : pfgate.gates) {
+        fgate.gates.push_back(gate);
+    }
+
+    for (auto link : pfgate.links) {
+      if (link->prev == nullptr) continue;
+
+      auto pgate = link->prev->val;
+
+      if (pgate->visited == 1) {
+        MakeGateSequence(max_fused_size, scratch, *pgate);
+      }
+    }
+  }
+
+  static void FindLongestGateSequence(
```

Comment:
After reading this, I think I understand the diagrams above a bit better. My reading is that this will probe both forwards and backwards in time from `fgate` to find the longest gate sequence (on no more than `max_fused_size` qubits) that it can be a part of. Is that correct?

Reply:
Yes, that is correct.